client/src/nat.rs

```diff
@@ -110,19 +112,37 @@ impl<'a> NatTraverse<'a> {
             .add_peers(&updates)
             .apply(self.interface, self.backend)?;
 
+        let mut peers_to_reset = vec![];
         let start = Instant::now();
         while start.elapsed() < STEP_INTERVAL {
-            self.refresh_remaining()?;
+            let mut exhausted = self.refresh_remaining()?;
+            peers_to_reset.append(&mut exhausted);
+
             if self.is_finished() {
                 log::debug!("NAT traverser is finished!");
                 break;
             }
             std::thread::sleep(Duration::from_millis(100));
         }
+
+        // For any peers which didn't connect and have no more candidates
+        // to try, reset their endpoint to the server-provided endpoint.
+        let reset_updates: Vec<_> = peer_reset_updates(peers_to_reset).collect();
+
+        DeviceUpdate::new()
+            .add_peers(&reset_updates)
+            .apply(self.interface, self.backend)?;
+
         Ok(())
     }
 }
 
+fn peer_reset_updates(peers_to_reset: Vec<Peer>) -> impl Iterator<Item = PeerConfigBuilder> {
```

Comment:
(if this function stays): Perhaps this could be called `reset_endpoints()`?

`peer_reset_updates()` sounds like "reset updates of a peer", but this operates on multiple peers, and sets them to server-reported endpoint. (it is somewhat ambiguous what "update" means here).

I see the name is complicated by the fact this does not actually do the change, rather provides the iterator of commands to do that.